src/components/HelpScoutBeacon/HelpScoutBeacon.js

```diff
@@ -0,0 +1,377 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import styled from 'styled-components'
+import { Transition, animated } from 'react-spring'
+import {
+  Button,
+  IconClose,
+  SafeLink,
+  breakpoint,
+  springs,
+  theme,
+} from '@aragon/ui'
+import BeaconHeadScripts from './BeaconHeadScripts'
+import IconQuestion from './IconQuestion'
+import logo from './logo.png'
+import { GU } from '../../utils'
+
+const HELPSCOUT_BEACON_KEY = 'helpscout-beacon'
+const CLOSED = 'closed, user can open opt-in dialogue'
+const OPENED = 'opened, user can opt-in or close'
+const OPENING = 'opening'
+const CLOSING = 'closing'
+
+const Beacon = React.memo(() => {
+  const [optedIn, setOptedIn] = React.useState(false)
+  React.useEffect(() => {
+    setOptedIn(localStorage.getItem(HELPSCOUT_BEACON_KEY) === '1')
+  }, [])
+  const handleOptIn = React.useCallback(() => {
+    localStorage.setItem(HELPSCOUT_BEACON_KEY, '1')
+    setOptedIn(true)
+  })
+
+  return (
+    <div
+      css={`
+        position: absolute;
+        bottom: ${2 * GU}px;
+        right: ${2 * GU}px;
+        z-index: 1001;
+
+        ${breakpoint(
+          'medium',
+          `
+            z-index: 10000;
+            bottom: ${3 * GU}px;
+            right: ${3 * GU}px;
+          `
+        )}
+      `}
+    >
+      <BeaconHeadScripts optedIn={optedIn} />
+      <HelpOptIn onOptIn={handleOptIn} optedIn={optedIn} />
+    </div>
+  )
+})
+
+const HelpOptIn = React.memo(({ onOptIn, optedIn }) => {
+  const [mode, setMode] = React.useState(CLOSED)
+  const handleClose = React.useCallback(() => setMode(CLOSED))
+  const handleToggle = React.useCallback(() => {
+    if (mode !== OPENING && mode !== CLOSING) {
+      setMode(mode === CLOSED ? OPENING : CLOSING)
+    }
+    if (optedIn) {
+      window.Beacon('toggle')
+    }
+  })
+  const handleToggleEnd = React.useCallback(() => {
+    setMode(mode === OPENING ? OPENED : CLOSED)
+  })
+  const handleOptIn = React.useCallback(() => {
+    onOptIn()
+    setTimeout(() => window.Beacon('open'), 100)
+  })
+  React.useEffect(() => {
+    if (optedIn && window.Beacon) {
+      window.Beacon('on', 'open', () => setMode(OPENED))
+      window.Beacon('on', 'close', () => setMode(CLOSED))
+    }
+  }, [optedIn, window.Beacon])
```

Comment:
Turns out, this is needed as `window.Beacon` changes from `undefined` to `function`

Reply:
mmm but since it’s not part of any React state, it will refresh the `useEffect()` at some point (because `window.Beacon` has changed), but the re-render responsible of it might be triggered by something else.

For example:

We could have `optedIn` changing from `false` to `true`, which would execute the `useEffect()` function with the new context, but `window.Beacon` would be `undefined` so it won’t run.

Then, `window.Beacon` would be set, but nothing will rerender `HelpScoutBeacon` so the `useEffect()` function won’t be executed again.

Then, at some point, something will re-render `HelpScoutBeacon`, and `window.Beacon` would finally be seen and the function re-executed. The problem is that the thing re-rendering `HelpScoutBeacon` is hard to predict if we don’t set it explicitly. Ideally, we should re-render when `window.Beacon` is ready… but I’m not sure how to know that?